contracts/external/cw-payroll-factory/src/contract.rs

```diff
@@ -0,0 +1,183 @@
+    info: MessageInfo,
+    instantiate_msg: PayrollInstantiateMsg,
+    code_id: u64,
+    label: String,
+) -> Result<Response, ContractError> {
+    // Instantiate the specified contract with owner as the admin.
+    let instantiate = WasmMsg::Instantiate {
+        admin: instantiate_msg.owner.clone(),
+        code_id,
+        msg: to_binary(&instantiate_msg)?,
+        funds: info.funds,
+        label,
+    };
+
+    let msg = SubMsg::reply_on_success(instantiate, INSTANTIATE_CONTRACT_REPLY_ID);
+
+    let instantiator = deps
+        .api
+        .addr_validate(&instantiate_msg.owner.unwrap_or(info.sender.to_string()))?;
+    let recipient = deps.api.addr_validate(&instantiate_msg.params.recipient)?;
+
+    // Save tmp contract info for use in reply
+    TMP_CONTRACT_INFO.save(deps.storage, &(recipient, instantiator))?;
+
+    Ok(Response::default()
+        .add_attribute("action", "instantiate_cw_payroll")
+        .add_submessage(msg))
+}
+
+#[cfg_attr(not(feature = "library"), entry_point)]
+pub fn query(deps: Deps, _env: Env, msg: QueryMsg) -> StdResult<Binary> {
+    match msg {
+        QueryMsg::ListVestingContracts { start_after, limit } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_after.as_deref().map(Bound::exclusive);
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .range(deps.storage, start, None, Order::Ascending)
+                .take(limit)
+                .map(|vc| Ok::<VestingContract, ContractError>(vc?.1))
+                .flatten()
+                .collect();
+
+            Ok(to_binary(&res)?)
+        }
+        QueryMsg::ListVestingContractsByInstantiator {
+            instantiator,
+            start_after,
+            limit,
+        } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_after.map(Bound::<String>::exclusive);
+
+            // Validate owner address
+            deps.api.addr_validate(&instantiator)?;
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .idx
+                .owner
+                .prefix(instantiator)
+                .range(deps.storage, start, None, Order::Ascending)
+                .take(limit)
+                .map(|vc| Ok::<VestingContract, ContractError>(vc?.1))
+                .flatten()
+                .collect();
+
+            Ok(to_binary(&res)?)
+        }
+        QueryMsg::ListVestingContractsByRecipient {
+            recipient,
+            start_after,
+            limit,
+        } => {
+            let limit = limit.unwrap_or(DEFAULT_LIMIT).min(MAX_LIMIT) as usize;
+            let start = start_after.map(Bound::<String>::exclusive);
+
+            // Validate recipient address
+            deps.api.addr_validate(&recipient)?;
+
+            let res: Vec<VestingContract> = vesting_contracts()
+                .idx
+                .recipient
```

Comment:
why did you decide to make one vesting contract per recipient?

Reply:
Simplifies a lot of things with regard to staking. Also minimizes risk of cross-contaminating funds. Note: I actually tried to make a multi-recipient vesting contract, but it was starting to get messy. There are a bunch of weird edge case around unbonding periods as well.

tl;dr thought it was simpler and safer.